lib/pbench/server/filetree.py

```diff
@@ -0,0 +1,883 @@
+
+        This is almost a mirror of the INCOMING tree structure, except that,
+        instead of directories containing unpacked tarballs, the RESULTS tree
+        contains only a symlink for each dataset name to the INCOMING tree's
+        unpacked tarball directory.
+
+        The link is created when each tarball is unpacked, and gives the normal
+        reference path for unpacked results data (usually through the link at
+        /var/www/html/results).
+
+    NOTE: this overview is specific to Pbench 0.72. While the descriptions here
+    are mostly applicable to 0.69 as well, there are additional complications
+    (and another root directory tree) we're dropping for 0.72, including the
+    entire USERS tree, which is replaced by real user ownership, and a "prefix"
+    mechanism that allowed tarballs to be nested within an arbitrary directory
+    hierarchy under the RESULTS controller directory.
+    """
+
+    # The FileTree class owns the definition of the "controller" level
+    # directory where PUT will store uploading files. Co-locating this
+    # with the ARCHIVE tree ensures that we can move files without an
+    # additional copy, and the upload will already fail if the file
+    # system has insufficient space. Defining the directory here allows
+    # FileTree discovery to ignore it.
+    TEMPORARY = "UPLOAD"
+
+    @staticmethod
+    def delete_if_empty(directory: Path) -> None:
+        """
+        Delete a directory only if it exists and is empty.
+
+        NOTE: rmdir technically will fail if the directory isn't empty, but
+        this feels safer.
+
+        Any exceptions raised will be propagated.
+
+        Args:
+            directory: Directory path
+        """
+        if directory.exists() and not any(directory.iterdir()):
+            directory.rmdir()
+
+    def __init__(self, options: PbenchServerConfig, logger: Logger):
+        """
+        Construct a FileTree object. We don't do any discovery here, because
+        the mutation operations allow dynamic minimal discovery to save on
+        some time. The `full_discovery` method allows full discovery when
+        desired.
+
+        Args:
+            options: PbenchServerConfig configuration object
+            logger: A Pbench python Logger
+        """
+        self.options = options
+        self.archive_root = self.options.ARCHIVE
+        self.incoming_root = self.options.INCOMING
+        self.results_root = self.options.RESULTS
+        self.logger = logger
+        self.controllers: Dict[str, Controller] = {}
+        self.datasets: Dict[str, Tarball] = {}
+
+    def full_discovery(self):
+        """
+        We discover the ARCHIVE, INCOMING, and RESULTS trees as defined by the
+        pbench-server.cfg file.
+
+        Full discovery is not required before adding or deleting a dataset.
+        """
+        self._discover_archive()
+        self._discover_unpacked()
+        self._discover_results()
```

Comment:
I think there's an ordering requirement here, which should probably be noted in the docstring or a comment.  (We have to discover all the controllers first, so we need to do the archive before we do the others.)